File: app/src/main/java/ru/devsp/app/mtgcollections/repository/CardLocalRepository.java
package ru.devsp.app.mtgcollections.repository;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import ru.devsp.app.mtgcollections.model.db.CardDao;
import ru.devsp.app.mtgcollections.model.objects.Card;
import ru.devsp.app.mtgcollections.model.objects.CardColorState;
import ru.devsp.app.mtgcollections.model.objects.CardExists;
import ru.devsp.app.mtgcollections.model.objects.CardForLibrary;
import ru.devsp.app.mtgcollections.model.objects.CardLibraryInfo;
import ru.devsp.app.mtgcollections.model.objects.CardManaState;
import ru.devsp.app.mtgcollections.model.objects.Color;
import ru.devsp.app.mtgcollections.model.objects.Filter;
import ru.devsp.app.mtgcollections.model.objects.SetName;
import ru.devsp.app.mtgcollections.model.objects.Subtype;
import ru.devsp.app.mtgcollections.model.objects.Type;
import ru.devsp.app.mtgcollections.tools.AppExecutors;

/**
 * Обработка сохраненных карт
 * Created by gen on 03.10.2017.
 */
@Singleton
public class CardLocalRepository {

    private final AppExecutors mAppExecutors;

    private CardDao mDao;

    @Inject
    CardLocalRepository(AppExecutors appExecutors, CardDao dao) {
        mAppExecutors = appExecutors;
        mDao = dao;
    }

    public void save(Card card) {
        mAppExecutors.diskIO().execute(() -> mDao.insert(card));
    }

    public void update(Card card) {
        mAppExecutors.diskIO().execute(() -> mDao.update(card));
    }

    public void updateLink(String parent, String link) {
        mAppExecutors.diskIO().execute(() -> mDao.updateLink(parent, link));
    }

    public void setChild(String card) {
        mAppExecutors.diskIO().execute(() -> mDao.updateChildState(card, true));
    }

    public LiveData<List<Card>> getFilteredCard(Filter filter) {
        return mDao.getFilteredCards(filter.types, filter.subtypes, filter.colors,
                filter.rarities, filter.sets);
    }

    /**
     * Все карты
     *
     * @return ld
     */
    public LiveData<List<Card>> getAllCards() {
        return mDao.getAllCards();
    }

    /**
     * Получение карт по колоде
     *
     * @param library library
     * @return ld
     */
    public LiveData<List<CardForLibrary>> getCardsByLibrary(long library) {
        return mDao.getCardsByLibrary(library);
    }

    /**
     * Список желаемого
     * @return ld
     */
    public LiveData<List<Card>> getWishList(){
        return mDao.getWishList();
    }

    public LiveData<Card> getCard(String id) {
        return mDao.getCard(id);
    }

    public LiveData<CardExists> getCardExists(String id) {
        return mDao.getCardExists(id);
    }

    public LiveData<List<Card>> getCardsBySet(String set){
        return mDao.getCardsBySet(set);
    }

    /**
     * Получение колод, в которые входит карта
     *
     * @param id id
     * @return ld
     */
    public LiveData<List<CardLibraryInfo>> getLibrariesByCard(String id) {
        return mDao.getLibrariesByCard(id);
    }

    public LiveData<List<CardManaState>> getLibraryManaState(long library) {
        return mDao.getLibraryManaState(library);
    }

    public LiveData<List<CardColorState>> getLibraryColorState(long library) {
        return mDao.getLibraryColorState(library);
    }

    private void getFilterParams(MutableLiveData<Filter> filterLiveData) {
        mAppExecutors.diskIO().execute(() -> {
            Filter filter = new Filter();

            filter.colors = getColors();
            filter.subtypes = getSubtypes();
            filter.rarities = getRarities();
            filter.sets = getSetNames();
            filter.types = getTypes();

            filterLiveData.postValue(filter);
        });
    }

    private String[] getRarities() {
        String[] strings = new String[6];
        strings[0] = "Basic Land";
        strings[1] = "Common";
        strings[2] = "Uncommon";
        strings[3] = "Rare";
        strings[4] = "Mythic Rare";
        strings[5] = "Special";
        return strings;
    }

    private String[] getTypes() {
        List<Type> list = mDao.getTypes();
        String[] strings = new String[list.size()];
        for (int i = 0; i < list.size(); i++) {
            strings[i] = list.get(i).type;
        }
        return strings;
    }

    private String[] getSubtypes() {
        List<Subtype> subtypes = mDao.getSubtypes();
        String[] strings = new String[subtypes.size()];
        for (int i = 0; i < subtypes.size(); i++) {
            strings[i] = subtypes.get(i).subtype;
        }
        return strings;
    }


    /**
     * Список сетов
     * @return Список сетов
     */
    private String[] getSetNames() {
        List<SetName> setNames = mDao.getSetNames();
        String[] strings = new String[setNames.size()];
        for (int i = 0; i < setNames.size(); i++) {
            strings[i] = setNames.get(i).setName;
        }
        return strings;
    }

    /**
     * Список цветов
     * @return Список цветов
     */
    private String[] getColors() {
        List<Color> colors = mDao.getColors();
        String[] strings = new String[colors.size()];
        for (int i = 0; i < colors.size(); i++) {
            strings[i] = colors.get(i).color;
        }
        return strings;
    }

    public LiveData<Filter> getFilter() {
        MutableLiveData<Filter> filter = new MutableLiveData<>();
        getFilterParams(filter);
        return filter;
    }

}

File: app/src/main/java/ru/devsp/app/mtgcollections/repository/CardLocalRepository.kt
package ru.devsp.app.mtgcollections.repository


import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData

import javax.inject.Inject
import javax.inject.Singleton

import ru.devsp.app.mtgcollections.model.db.CardDao
import ru.devsp.app.mtgcollections.model.objects.Card
import ru.devsp.app.mtgcollections.model.objects.CardColorState
import ru.devsp.app.mtgcollections.model.objects.CardExists
import ru.devsp.app.mtgcollections.model.objects.CardForLibrary
import ru.devsp.app.mtgcollections.model.objects.CardLibraryInfo
import ru.devsp.app.mtgcollections.model.objects.CardManaState
import ru.devsp.app.mtgcollections.model.objects.Filter
import ru.devsp.app.mtgcollections.tools.AppExecutors

/**
 * Обработка сохраненных карт
 * Created by gen on 03.10.2017.
 */
@Singleton
class CardLocalRepository @Inject
internal constructor(private val appExecutors: AppExecutors, private val dao: CardDao) {

    /**
     * Все карты
     *
     * @return ld
     */
    val allCards: LiveData<List<Card>>
        get() = dao.allCards

    /**
     * Список желаемого
     * @return ld
     */
    val wishList: LiveData<List<Card>>
        get() = dao.wishList

    private val rarities: Array<String>
        get() {
            return arrayOf(
                    "Basic Land",
                    "Common",
                    "Uncommon",
                    "Rare",
                    "Mythic Rare",
                    "Special"
            )
        }

    private val types: Array<String>
        get() {
            val list = dao.types
            return Array(list.size, { i -> list[i].type })
        }

    private val subtypes: Array<String>
        get() {
            val subtypes = dao.subtypes
            return Array(subtypes.size, { i -> subtypes[i].subtype })
        }

    /**
     * Список сетов
     * @return Список сетов
     */
    private val setNames: Array<String>
        get() {
            val setNames = dao.setNames
            return Array(setNames.size, { i -> setNames[i].setName })
        }

    /**
     * Список цветов
     * @return Список цветов
     */
    private val colors: Array<String>
        get() {
            val colors = dao.colors
            return Array(colors.size, { i -> colors[i].color })
        }

    val filter: LiveData<Filter>
        get() {
            val filter = MutableLiveData<Filter>()
            getFilterParams(filter)
            return filter
        }

    fun save(card: Card) {
        appExecutors.diskIO().execute { dao.insert(card) }
    }

    fun update(card: Card) {
        appExecutors.diskIO().execute { dao.update(card) }
    }

    fun updateLink(parent: String, link: String) {
        appExecutors.diskIO().execute { dao.updateLink(parent, link) }
    }

    fun setChild(card: String) {
        appExecutors.diskIO().execute { dao.updateChildState(card, true) }
    }

    fun getFilteredCard(filter: Filter): LiveData<List<Card>> {
        return dao.getFilteredCards(filter.types, filter.subtypes, filter.colors,
                filter.rarities, filter.sets)
    }

    /**
     * Получение карт по колоде
     *
     * @param library library
     * @return ld
     */
    fun getCardsByLibrary(library: Long): LiveData<List<CardForLibrary>> {
        return dao.getCardsByLibrary(library)
    }

    fun getCard(id: String): LiveData<Card> {
        return dao.getCard(id)
    }

    fun getCardExists(id: String): LiveData<CardExists> {
        return dao.getCardExists(id)
    }

    fun getCardsBySet(set: String): LiveData<List<Card>> {
        return dao.getCardsBySet(set)
    }

    /**
     * Получение колод, в которые входит карта
     *
     * @param id id
     * @return ld
     */
    fun getLibrariesByCard(id: String): LiveData<List<CardLibraryInfo>> {
        return dao.getLibrariesByCard(id)
    }

    fun getLibraryManaState(library: Long): LiveData<List<CardManaState>> {
        return dao.getLibraryManaState(library)
    }

    fun getLibraryColorState(library: Long): LiveData<List<CardColorState>> {
        return dao.getLibraryColorState(library)
    }

    private fun getFilterParams(filterLiveData: MutableLiveData<Filter>) {
        appExecutors.diskIO().execute {
            val filter = Filter()

            filter.colors = colors
            filter.subtypes = subtypes
            filter.rarities = rarities
            filter.sets = setNames
            filter.types = types

            filterLiveData.postValue(filter)
        }
    }

}
